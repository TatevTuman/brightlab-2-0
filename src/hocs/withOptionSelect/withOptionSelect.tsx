import React, { useEffect, useState } from 'react'
import { OptionType } from '@types'

export type WithOptionSelectProps = {
  value: string | null
  options: OptionType[]
  onChange: (value: string | null) => void
  defaultValue?: string
}

export type WithOptionSelectState = {
  selectedOption: OptionType | null
  options: OptionType[]
}

export type WithOptionSelectPropsPassed = {
  options: WithOptionSelectProps['options']
  selectedOption: WithOptionSelectState['selectedOption']
  handleOptionSelect: (option: OptionType | null) => void
}

const withOptionSelect = <P,>(WrappedComponent: React.ComponentType<WithOptionSelectPropsPassed & P>) => {
  const OptionSelect: React.FC<WithOptionSelectProps & P> = props => {
    const { value, defaultValue, options, onChange, ...parentProps } = props

    const [state, changeState] = useState<WithOptionSelectState>({
      selectedOption: null,
      options: []
    })

    /* Like React.Component setState function */
    const setState = (nextState: Partial<WithOptionSelectState>) => {
      changeState({ ...state, ...nextState })
    }

    /*
      Sets selectedOption from props defaultValue
      Sets options from props
    */
    useEffect(() => {
      if (defaultValue) {
        const defaultSelectedOption = options.find(option => option.value === defaultValue)
        setState({ selectedOption: defaultSelectedOption })
      }

      setState({ options })
    }, [])

    /* Set selectedOption from props value */
    useEffect(() => {
      const nextSelectedOption = options.find(option => option.value === value)
      setState({ selectedOption: nextSelectedOption })
    }, [value])

    /* Handles option selecting and calls props.onSelect */
    const handleOptionSelect = (option: OptionType | null) => {
      setState({ selectedOption: option })

      const optionValue = option && option.value
      onChange && onChange(optionValue)
    }

    return (
      <WrappedComponent
        options={options}
        selectedOption={state.selectedOption}
        handleOptionSelect={handleOptionSelect}
        {...(parentProps as P)}
      />
    )
  }

  OptionSelect.defaultProps = ({
    value: null,
    options: [],
    onSelect: () => null
  } as unknown) as WithOptionSelectProps & P

  return OptionSelect
}

export default withOptionSelect
