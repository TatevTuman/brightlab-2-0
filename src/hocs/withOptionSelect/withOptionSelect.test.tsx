import React from 'react'
import { render, cleanup, waitFor, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import withOptionSelect from './withOptionSelect'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('withOptionSelect', () => {
  const WithOptionSelectComponent = withOptionSelect(props => {
    const { selectedOption, handleOptionSelect, options } = props

    return (
      <div data-testid={'with-option-select-component'}>
        <div>{selectedOption?.value}</div>
        {options.map(option => {
          return (
            <div
              key={option.value}
              className={'with-option-select-component-option'}
              onClick={() => handleOptionSelect(option)}
            >
              {option.label}
            </div>
          )
        })}
      </div>
    )
  })

  const props = {
    options: [{ label: 'Test', value: 'test' }],
    onChange: jest.fn(),
    value: null
  }

  it('renders correctly', async () => {
    const { container } = render(<WithOptionSelectComponent {...props} />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
  })

  it('renders options correctly', async () => {
    const { container } = render(<WithOptionSelectComponent {...props} />)
    await waitFor(() => container)

    const options = container.querySelectorAll('.with-option-select-component-option')
    expect(options).toHaveLength(props.options.length)
  })

  it('renders value correctly', async () => {
    const { container, getByText } = render(<WithOptionSelectComponent {...props} value={props.options[0].value} />)
    await waitFor(() => container)

    expect(getByText(props.options[0].value)).toBeInTheDocument()
  })

  it('handles onChange', async () => {
    const { container, getByText } = render(<WithOptionSelectComponent {...props} />)
    await waitFor(() => container)

    const optionLabel = getByText(props.options[0].label)

    act(() => {
      userEvent.click(optionLabel)
    })

    const optionValue = getByText(props.options[0].value)

    expect(optionValue).toBeInTheDocument()
    expect(props.onChange).toHaveBeenCalledWith(props.options[0].value)
  })
})
