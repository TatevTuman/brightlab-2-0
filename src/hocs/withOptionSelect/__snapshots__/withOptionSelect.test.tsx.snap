// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withOptionSelect renders correctly 1`] = `
<div>
  <div
    data-testid="with-option-select-component"
  >
    <div />
    <div
      class="with-option-select-component-option"
    >
      Test
    </div>
  </div>
</div>
`;
